specs/ExecuteScript.md

```diff
@@ -0,0 +1,222 @@
+# Background
+Currently WebView2 supports script execution in the main page only -
+documentation can be found [here](https://docs.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/icorewebview2?view=webview2-1.0.902.49#executescript).
+With introduction of API for [separate iframes](https://docs.microsoft.com/en-us/microsoft-edge/webview2/reference/win32/icorewebview2frame?view=webview2-1.0.902.49) we find it useful to support script execution in child frames also.
+The time when script can be executed is tightly coupled with the navigation state of the page,
+therefore we additionally providing [navigation events](https://docs.microsoft.com/en-us/microsoft-edge/webview2/concepts/navigation-events) in the iframes - NavigationStarting, ContentLoading, NavigationCompleted and DOMContentLoaded event.
+
+In this document we describe API additions to support ExecuteScript and navigation
+events in iframes. We'd appreciate your feedback.
+
+
+# Description
+Using ExecuteScript user can run script in the context of a calling frame and receive the script result in the callback if provided.
+If ExecuteScript is run during the navigation it should be called after ContentLoading event.
+
+# Examples
+## DOMContentLoaded event and ExecuteScript
+DOMContentLoaded is raised when the initial HTML document has been parsed.
+
+## .NET, WinRT
+```c#
+    webView.CoreWebView2.FrameCreated += (sender, args) =>
+    {
+        args.Frame.DOMContentLoaded += (frameSender, DOMContentLoadedArgs) =>
+        {
+            args.Frame.ExecuteScriptAsync(
+                "let content = document.createElement(\"h2\");" +
+                "content.style.color = 'blue';" +
+                "content.textContent = \"The url of the iframe is \" + window.location.href.toString();" +
+                "document.body.appendChild(content);");
+        };
+    };
+```
+
+## Win32 C++
+```cpp
+    CHECK_FAILURE(m_appWindow->GetWebView()->QueryInterface(IID_PPV_ARGS(&m_webView2)));
+    wil::com_ptr<ICoreWebView2_4> webview2_4 = m_webView.try_query<ICoreWebView2_4>();
```

Comment:
For .NET & WinRT there's one class with all of the known APIs. If an API has no implementation because the SDK is newer than the runtime, the API will throw.